sdk/core/azure-core/inc/context.hpp

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <assert.h>
+#include <chrono>
+#include <memory>
+//#include <stdio.h>
+#include <stdlib.h>
+#include <string>
+#include <type_traits>
+
+constexpr char not_found = 0;
+
+namespace Azure
+{
+namespace Core
+{
+
+struct ValueBase
+{
+  virtual ~ValueBase() {}
+};
+
+/*
+ *  @brief ContextValue exists as a substitute for variant which isn't available until C++17
+ */
+class ContextValue
+{
+  char active;
+  union {
+    bool b; // if active == 1
+    int i; // if active == 2
+    std::string s; // if active == 3
+    std::unique_ptr<ValueBase> p; // if active == 4
+  };
+
+public:
+  ContextValue() noexcept : active(0) {}
+  ContextValue(bool b) noexcept : active(1), b(b) {}
+  ContextValue(int i) noexcept : active(2), i(i) {}
+  ContextValue(const std::string& s) : active(3), s(s) {}
+  ContextValue(std::string&& s) noexcept : active(3), s(std::move(s)) {}
+  template <
+      class DerivedFromValueBase,
+      typename std::enable_if<std::is_convertible<DerivedFromValueBase*, ValueBase*>::value, int>::
+          type
+      = 0>
+  ContextValue(std::unique_ptr<DerivedFromValueBase>&& p) noexcept : active(4), p(std::move(p))
+  {
+  }
+
+  ContextValue(ContextValue&& other) : active(other.active)
+  {
+    switch (active)
+    {
+      case 1:
+        b = other.b;
+        break;
+      case 2:
+        i = other.i;
+        break;
+      case 3:
+        ::new (&s) std::string(std::move(other.s));
+        break;
+      case 4:
+        ::new (&p) std::unique_ptr<ValueBase>(std::move(other.p));
+        break;
+    }
+  }
+
+  ~ContextValue()
+  {
+    switch (active)
+    {
+      case 3:
+        s.~basic_string();
+        break;
+      case 4:
+        p.~unique_ptr<ValueBase>();
+        break;
+    }
+  }
+
+  ContextValue& operator=(const ContextValue& other) = delete;
+
+  template <class ExpectedType> const ExpectedType& get() const noexcept;
+
+  char alternative() const noexcept { return active; }
+};
+
+template <> inline const bool& ContextValue::get() const noexcept
+{
+  if (active != 1)
+  {
+    abort();
+  }
+  return b;
+}
+
+template <> inline const int& ContextValue::get() const noexcept
+{
+  if (active != 2)
+  {
+    abort();
+  }
+  return i;
+}
+
+template <> inline const std::string& ContextValue::get() const noexcept
+{
+  if (active != 3)
+  {
+    abort();
+  }
+  return s;
+}
+
+template <> inline const std::unique_ptr<ValueBase>& ContextValue::get() const noexcept
+{
+  if (active != 4)
+  {
+    abort();
+  }
+  return p;
+}
+
+class Context
+{
+public:
+  using time_point = std::chrono::system_clock::time_point;
```

Comment:
This is still lowercase but it's attempting to follow the standard name. Should we put an exception for that into the guidelines?

Reply:
In this case I think it makes sense to stay like this.  If I were creating an instance of time_point then I would expect time_point m_timePoint.